File: JuLayout/OCLayout/JuLayout.m
//
//  JuLayout.m
//  JuLayout
//
//  Created by Juvid on 16/7/17.
//  Copyright © 2016年 Juvid. All rights reserved.
//

#import "JuLayout.h"
#import "UIView+JuLayout.h"

@implementation JuLayout
-(id)init{
    self=[super init];
    if (self) {
        self.juMulti=1;
        _prioritys=UILayoutPriorityRequired;
    }
    return self;
}
/**
 *  @author Juvid, 16-07-18 09:07:29
 *
 *  添加约束
 *
 *  @return 约束
 */
-(NSLayoutConstraint *)juAddConstraint{
     if (!_juView1.superview) return nil;
    _juView1.translatesAutoresizingMaskIntoConstraints=NO;
    UIView *toItem=_juView2;
    UIView *ju_View=_juView1.superview;
    CGFloat constant=_juConstant;
    if (self.juAttr1==NSLayoutAttributeWidth||_juAttr1==NSLayoutAttributeHeight) {
        if (!toItem) {
            if (constant==0) {///< 等于0时约束等于父view
                toItem=_juView1.superview;
            }else{///< 自己加约束
                ju_View=_juView1;
            }
        }
        else if([toItem isEqual:_juView1]){///< 长宽比例时
            ju_View=_juView1;
        }
    }else{
        if (!toItem) {
            toItem=_juView1.superview;
        }
    }
    NSLayoutConstraint *layoutConstraint=[NSLayoutConstraint constraintWithItem:_juView1 attribute:_juAttr1 relatedBy:_juRelation toItem:toItem attribute:_juAttr2 multiplier:_juMulti constant:constant];
    layoutConstraint.priority=_prioritys;
    layoutConstraint.juLayType=_juLayoutType;
    
    [_juView1 juCompareSameCons:layoutConstraint];
    
    if (!_juView1.ju_Constraints) _juView1.ju_Constraints=[NSMutableArray array];
    [_juView1.ju_Constraints addObject:layoutConstraint];
    
    [ju_View addConstraint:layoutConstraint];
    
    return layoutConstraint;
}
/***************************************************************/

-(void)setJuConstant:(CGFloat)juConstant{
    _juConstant=juConstant*(_isMinus?-1:1);
}
-(void)setJuLayoutType:(JuLayoutType)juLayoutType{
    _juLayoutType=juLayoutType;
    if (juLayoutType==JuLayoutAspectWH) {
        _juView2=_juView1;
    }
}
/***=====================添加过程约束==============================*/

-(JuLayout *(^)(CGFloat  prioritys))priority{
    return ^JuLayout* (CGFloat  prioritys){
        _prioritys=prioritys;
        return self;
    };
}
-(JuLayout *(^)(CGFloat  mulits))multi{
    return ^JuLayout* (CGFloat  mulits){
        _juMulti=mulits;
        return self;
    };
}
-(JuLayout *(^)(UIView * toItem))toView{
    return ^JuLayout* (UIView * toItem){///< block实现返回block变量
        _juView2=toItem;
        return self;///< block返回值
    };
}

/**======================最终生成约束=============================**/

-(NSLayoutConstraint *(^)(CGFloat constion))equal{
    return ^NSLayoutConstraint* (CGFloat constion){
         _juRelation=NSLayoutRelationEqual;
        self.juConstant=constion;
        return  [self juAddConstraint];
    };
}
-(NSLayoutConstraint *(^)(CGFloat constion))greatEqual{
    return ^NSLayoutConstraint* (CGFloat constion){
         _juRelation=NSLayoutRelationGreaterThanOrEqual;
        self.juConstant=constion;
        return [self juAddConstraint];
    };
}
-(NSLayoutConstraint *(^)(CGFloat constion))lessEqual{
    return ^NSLayoutConstraint* (CGFloat constion){
         _juRelation=NSLayoutRelationLessThanOrEqual;
        self.juConstant=constion;
        return [self juAddConstraint];
    };
}

@end

File: JuLayout/JuLayout.swift
//
//  JuLayout.swift
//  JuLayout
//
//  Created by Juvid on 2016/12/16.
//  Copyright © 2016年 Juvid. All rights reserved.
//

import UIKit




class JuLayout: NSObject {

    var isMinus :Bool = false ///< 只能为
    var juView2 : UIView?
    var juView1 : UIView!
    var juMulti : CGFloat = 1.0

    var juAttr1 : NSLayoutAttribute?
    var juAttr2 : NSLayoutAttribute?

    var juConstant : CGFloat = 1.0
    var juRelation  = NSLayoutRelation.equal
    var juPrioritys  = UILayoutPriorityRequired
    var juLayoutType :JuLayoutType?{
        didSet{
            if juLayoutType  == JuLayoutType.aspectWH {
                juView2 = juView1
            }
        }
    }


    func multi(_ mulits:CGFloat) -> JuLayout {
        juMulti = mulits
        return self;
    }

    func toView(_ toItem:UIView?) -> JuLayout {
        juView2 = toItem
        return self
    }
    func priority(_ prioritys:Float) -> JuLayout {
        juPrioritys = prioritys
        return self
    }
    func equal(_ constion:CGFloat)  {
        juRelation = NSLayoutRelation.equal;
        self.juAddConstraint(constion)
    }
    func greaterEqual(_ constion:CGFloat) {
        juRelation = NSLayoutRelation.greaterThanOrEqual;
        self.juAddConstraint(constion)
    }
    func lessEqual(_ constion:CGFloat)  {
         juRelation = NSLayoutRelation.lessThanOrEqual;
         self.juAddConstraint(constion)
    }
   
    func juAddConstraint(_ constions:CGFloat)  {
        if juView1.superview == nil {
            return
        }
        juConstant=constions*(isMinus ? -1 : 1)
        juView1.translatesAutoresizingMaskIntoConstraints = false
        var toItem = juView2
        var ju_View :UIView! = juView1.superview
        let constant = juConstant
        if (juAttr1 == NSLayoutAttribute.width || juAttr1 == NSLayoutAttribute.height ) {
            if toItem == nil {
                if constant == 0 {
                    toItem = juView1.superview
                }else{
                    ju_View = juView1
                }
            }
            else if (toItem?.isEqual(juView1))!{

            }
        }else{
            if toItem == nil {
                toItem = juView1.superview
            }
        }

        let layoutConstraint = NSLayoutConstraint.init(item: juView1, attribute: juAttr1!, relatedBy:juRelation, toItem: toItem, attribute: juAttr2!, multiplier:juMulti, constant: constant)
        layoutConstraint.priority = juPrioritys
        ju_View.addConstraint(layoutConstraint);
        layoutConstraint.juLayType=juLayoutType;
        
        juView1.juCompareSameCons(layoutConstraint)
        if juView1.ju_Constraints == nil {
            juView1.ju_Constraints = NSMutableArray.init()
        }
        juView1.ju_Constraints?.add(layoutConstraint)

    }

}
